pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -1796,21 +1799,37 @@ func getSharedDatastoresInClusters(ctx context.Context, clusterMorefs []string,
 func (c *K8sOrchestrator) StartZonesInformer(ctx context.Context,
 	restClientConfig *restclient.Config, namespace string) error {
 	log := logger.GetLogger(ctx)
+	var startErr error
+	startZoneInformerOnce.Do(func() {
+		var err error
+		if restClientConfig == nil {
+			restClientConfig, err = clientconfig.GetConfig()
+			if err != nil {
+				startErr = logger.LogNewErrorf(log, "failed to get restClientConfig. Error: %+v", err)
+				return
+			}
+		}
+		// Create an informer for Zone instances.
+		dynInformer, err := k8s.GetDynamicInformer(ctx, "topology.tanzu.vmware.com",
+			"v1alpha1", "zones", namespace, restClientConfig, false)
+		if err != nil {
+			startErr = logger.LogNewErrorf(log, "failed to create dynamic informer for Zones CR. Error: %+v", err)
+			return
+		}
+		zoneInformer = dynInformer.Informer()
 
-	// Create an informer for Zone instances.
-	dynInformer, err := k8s.GetDynamicInformer(ctx, "topology.tanzu.vmware.com",
-		"v1alpha1", "zones", namespace, restClientConfig, false)
-	if err != nil {
-		return logger.LogNewErrorf(log, "failed to create dynamic informer for Zones CR. Error: %+v", err)
-	}
-	zoneInformer = dynInformer.Informer()
-
-	// Start informer.
-	go func() {
-		log.Info("Informer to watch on Zones CR starting..")
-		zoneInformer.Run(make(chan struct{}))
-	}()
-	return nil
+		// Start informer in goroutine
+		go func() {
+			log.Info("Informer to watch on Zones CR starting..")
+			zoneInformer.Run(make(chan struct{}))
+		}()
+		// Wait for informer cache to sync
+		if !cache.WaitForCacheSync(ctx.Done(), zoneInformer.HasSynced) {
+			startErr = logger.LogNewErrorf(log, "Zone informer cache sync failed")
+			return
+		}
```

Comment:
If there is error here, what happens to the go routine above?

Reply:
refactored code to handle case of stopping routine and resetting startZoneInformerOnce for the case caller want to retry calling StartZonesInformer again.